### PR TITLE
[compiled autograd] torch.compile API

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125880

- enter existing compiled autograd ctx manager before entering torch.compile frames

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/__init__.py [hunk 1]
```diff
@@ -1898,14 +1901,24 @@ def fn(model: Callable):
...
 
+    def rewrap(model):
+        return functools.partial(
+            compile,
+            fullgraph=fullgraph,
+            dynamic=dynamic,
+            backend=backend,
+            mode=mode,
+            options=options,
+            disable=disable)(model)
```
Comment (jansel) [inline torch/__init__.py:1912]: I'm a bit confused by the usage of partial here, perhaps you want:
```suggestion
    rewrap = functools.partial(
            compile,
            fullgraph=fullgraph,
            dynamic=dynamic,
            backend=backend,
            mode=mode,
            options=options,
            disable=disable)
```

File: torch/_dynamo/compiled_autograd.py [hunk 2]
```diff
@@ -269,35 +286,59 @@ def set_node_origin(self, node_name, node_index):
...
+
+
+# Forces .backward() call within to dispatch to compiled autograd
+# Allows for more fine-grained control over which autograd engine calls should be handled
 @contextlib.contextmanager
 def enable(compiler_fn):
-    global compiled_autograd_enabled, override_compiler_fn
-    assert not compiled_autograd_enabled
-    compiled_autograd_enabled = True
+    global override_compiler_fn
```
Comment (jansel) [inline torch/_dynamo/compiled_autograd.py:314]: Why is this called "override" are we overriding some other setting?  Or is this just the main way to set it?

File: torch/_dynamo/config.py [hunk 3]
```diff
@@ -437,6 +437,11 @@ def default_debug_dir_root():
 # WARNING: this is an experimental flag and is subject to change.
 _experimental_support_context_fn_in_torch_utils_checkpoint = False
 
+# This flag is used during forward compilation to prepare compile modified autograd nodes
+# for compiled autograd. To disable compiled autograd after the forward compilation, use the
+# torch._dynamo.compiled_autograd.disable() context manager instead
+compiled_autograd_enabled = False
```
Comment (jansel) [inline torch/_dynamo/config.py:443]:
```suggestion
compiled_autograd = False
```

File: torch/_dynamo/eval_frame.py [hunk 4]
```diff
@@ -668,6 +668,8 @@ def toy_example(a, b):
     torch._C._log_api_usage_once("torch._dynamo.optimize")
     if disable or os.environ.get("TORCHDYNAMO_DISABLE", "") == "1":
         return _NullDecorator()
+    elif config.compiled_autograd_enabled:
+        torch._dynamo.compiled_autograd._initialize()
```
Comment (jansel) [inline torch/_dynamo/eval_frame.py:672]: Wont installing compiled autograd globally like this has bad global side effects?

If I have two models, one compiled and one eager -- then this will enable compiled autograd for the unrelated model.

**Review by jansel — CHANGES_REQUESTED** [review]
Test failures

File: torch/_dynamo/compiled_autograd.py (hunk 2 above)
Comment (xmfan) [reply torch/_dynamo/compiled_autograd.py:314]: suppose we have 
```python
torch._dynamo.config.compiled_autograd = True
compiled_fn = torch.compile(fn, backend="inductor")  # _initialize()
loss = compiled_fn()
with enable(torch.compile(backend="aot_eager")):
  loss.backward()
```

backend aot_eager overrides the inductor compile wrapper set on the autograd nodes

File: torch/_dynamo/eval_frame.py (hunk 4 above)
Comment (xmfan) [reply torch/_dynamo/eval_frame.py:672]: I don't think it should? I changed the autograd engine routing logic in engine.cpp:1270. it now needs
- a compiled being set (by _initialize)
- either
  - an autograd graph containing a CompiledFunctionBackward marked by compiled autograd
  - .backward called within the ctx manager override

the eager model's output on which we call .backward shouldn't have an autograd graph containing CompiledFunctionBackward, so autograd engine would route to eager.

**Review by albanD — COMMENTED** [review]
I'm a bit surprised that torch.compile now has effect at a distance where code unrelated to the fw/bw it captures will also run under compile?
I was under the expectation that only .backward() calls within torch.compile would trigger compiled autograd?
Can you share some details on why this has changed?

**Comment by xmfan** [comment]
@albanD Could you elaborate more? If some tensor's autograd graph contains a node created by torch.compile (e.g. CompiledFunctionBackward), it should run compiled autograd.

The context manager API that would let users to do eager fw + CA bw is there mostly for tests

**Comment by albanD** [comment]
> If some tensor's autograd graph contains a node created by torch.compile (e.g. CompiledFunctionBackward), it should run compiled autograd.

My question is why? This custom Function already properly contains the compiled backward for the op that was marked for compilation during the forward.
But enabling CA overall here would trigger compilation for the backward of ops that were not in the compiled region in the forward right?

**Comment by xmfan** [comment]
Yes, CA tries to compile the entire backward, even the nodes originating from fw graph break regions. Without this, we would have equal sized graphs as AOTAutograd, and can't capture DDP/FSDP hooks

**Comment by albanD** [comment]
> If I understand the concern properly, CA should only handle gradients for nodes originating between the compiled start and compiled end regions in this example:

My mental model of what happens here is that we only ever compile code that is explicitly wrapped via torch.compile.
So doing torch.compile of a function in the forward will compile that function and the corresponding backward.
If the user wants the whole backward to be compiled, they should wrap the call to .backward() within torch.compile to state that they want all of it to be compiled.

I find the current logic that will trigger compiled backward beyond what is torch.compile-d explicitly surprising. And I feel like a lot of the complexity of this PR comes from this: We have to leave breadcrumbs in the autograd.Function generated by AOTAutograd so that we can then traverse the whole autograd graph and find such a Node to then trigger a full compile.
This sounds very complex and brittle compared to `torch.compile(lambda: loss.backward())()` that I would expect the user to do if they want their whole backward to be compiled in a single block.

**Comment by xmfan** [comment]
That makes a lot of sense to me, I'll explore the implementation

**Comment by xmfan** [comment]
@jansel could you take another look

**Review by jansel — APPROVED** [review]
